### PR TITLE
Add Types to Redux Store

### PR DESCRIPTION
**Pull request by alecmerdler**
### Description

Heavily inspired by [this TypeScript/Redux/React guide](https://github.com/piotrwitek/react-redux-typescript-guide).

### Changes

- Introduces [`typesafe-actions`](https://github.com/piotrwitek/typesafe-actions) library to improve Redux usage
- Move Redux modules to `actions/` and `reducers/` directories
- Remove `redux-form` dependency in favor of React hooks (only 1 usage)
- Implement `withHandlePromise` [HOC](https://reactjs.org/docs/higher-order-components.html) (to eventually replace `extends PromiseComponent`)

### Future Work

- Use `withHandlePromise` instead of `extends PromiseComponent`
- Further refine Redux action/reducer types
- Ensure all uses of `connect()` are type-safe when accessing the Redux store

Addresses https://jira.coreos.com/browse/CONSOLE-1412

### CODE REVIEW

**Comment by spadgett** [comment]
cc @vojtechszocs @christianvogt

**Comment by spadgett** [comment]
```
ERROR in /go/src/github.com/openshift/console/frontend/public/actions/ui.ts
(154,26): Argument of type 'Dispatch<AnyAction>' is not assignable to parameter of type 'ObjectIterator<((dispatch: any) => Promise<void | { type: ActionType; flag: FLAGS; value: boolean; } | { type: ActionType.Resources; resources: { models: K8sKind[]; }; }>)[], any>'.
  Types of parameters 'action' and 'value' are incompatible.
    Type 'number | ((dispatch: any) => Promise<void | { type: ActionType; flag: FLAGS; value: boolean; } | { type: ActionType.Resources; resources: { models: K8sKind[]; }; }>) | (() => string) | { (...items: ConcatArray<(dispatch: any) => Promise<...>>[]): ((dispatch: any) => Promise<...>)[]; (...items: (((dispatch: any) => P...' is not assignable to type 'AnyAction'.
      Type 'number' is not assignable to type 'AnyAction'.

ERROR in /go/src/github.com/openshift/console/frontend/public/actions/ui.ts
(159,26): Argument of type 'Dispatch<AnyAction>' is not assignable to parameter of type 'ObjectIterator<((dispatch: any) => Promise<void | { type: ActionType; flag: FLAGS; value: boolean; } | { type: ActionType.Resources; resources: { models: K8sKind[]; }; }>)[], any>'.
```

File: frontend/public/components/utils/promise-component.tsx [hunk 1]
```diff
@@ -2,6 +2,31 @@
 
 import * as React from 'react';
 
+// TODO(alecmerdler): Write HOC for handling promise
```
Comment (spadgett) [inline frontend/public/components/utils/promise-component.tsx]: TODO still valid? I see `WithHandlePromise` below

**Comment by christianvogt** [comment]
@spadgett @alecmerdler 
I just skimmed the general approach to types and it's a good step in the right direction.

My preference is to create feature folders in the redux folder so that the actions, selectors, and reducer for a particular state are colocated together. As oppose to separate actions and reducers folders.

I did notice that some actions are still not typesafe when used in reducers.
We did some work in this area in OSIO to have typesafe redux. Perhaps we can draw some inspiration from the work done there: https://github.com/fabric8-ui/fabric8-ui/tree/master/packages/toolchain/src/app/redux

**Comment by alecmerdler** [comment]
@christianvogt Thanks for the feedback. I'll check out the code you linked. I'm not against having feature folders, the only thing is we have a couple actions that have multiple reducers (`K8sActionTypes.Resources`), but those could easily be handled. I'm still working on this, so the type safety of the reducers should improve. I'll keep you notified. The goal of this work is to make us more Redux-friendly and move away from bulky stateful classes (which are hard to reason about and cause bugs).

**Comment by alecmerdler** [comment]
@spadgett Everything is green, ready for another review.

**Pull request merged**